### DISCUSSION TITLE
Is array of user defined types forbidden for good?

### DISCUSSION THREAD

**Discussion by isidentical** [geldata/gel#2144] [Q&A]
https://github.com/edgedb/edgedb/blob/b56cb59ec6babf2f0437cfacc42a8804ab33201f/tests/test_schema.py#L161-L172

I was wondering whether this is just for the alpha, or embedded in the language spec? Since my current workflow includes creating a set, and during query converting to an array with `array_agg` though I think this is somewhat inefficient compared to storing that set of items as an array in the first place.

**Comment by elprans** [comment]
The trouble with arrays (and other collection types) is that it is very hard (or event impossible) to efficiently maintain referential integrity if you store object references in them.  Also, `array_agg` by itself isn't very expensive (although resolving the linked sets might be, depending on the cardinality of the source set).  

Is there any particular query you are having trouble with performance-wise?

**Reply by isidentical** [reply]
> The trouble with arrays (and other collection types) is that it is very hard (or event impossible) to efficiently maintain referential integrity if you store object references in them

Ah, I see. What about cases like where the array and any of its contents (or anything they point to) won't change at all (all references, everything is immutable)?

I also stumbled (I think) something relevant with this during my experiments while trying to create an array with array constructor (like if I do `[(SELECT type1 LIMIT 1), (SELECT type2 LIMIT 1)]` it doesn't allow me, throws an error regarding their size differ. But if they all extend from a common base, like `basetype`, then I can do this `[(SELECT basetype FILTER .id = <uuid>"test"), SELECT basetype FILTER .id = <uuid>"test2")]`, just a [possibly] irrelevant finding)

>  Also, array_agg by itself isn't very expensive (although resolving the linked sets might be, depending on the cardinality of the source set).

Depending on the input, it might be required to call `array_agg` thousands of times which I guesstimated would be inefficient (also checking out some other random resources on the internet, they claim `array_agg` on *postgres* is relatively slower). Though I can't claim anything without proper benchmarking/analysis (maybe possibly measured with #981), so I'm just speculating.

**Reply by elprans** [reply]
> while trying to create an array with array constructor (like if I do `[(SELECT type1 LIMIT 1), (SELECT type2 LIMIT 1)]` it doesn't allow me

Hm.  That might be a bug.  Please file an issue with an example query.
